release: v1.34.0 (#4207)
apollographql/router: CHANGELOG.md

Errors in responses may be:
 - moved to extensions, to avoid raising exceptions in clients
 - or disabled entirely, in which case clients will not receive any authorization errors.

By [@Geal](https://github.com/Geal) in https://github.com/apollographql/router/pull/4076 & https://github.com/apollographql/router/pull/4122

### Add a new studio reporting metric ([Issue #3883](https://github.com/apollographql/router/issues/3883))

Count how many reports we have submitted to studio, with a "type" attribute that indicates if reporting "traces" or "metrics".

By [@garypen](https://github.com/garypen) in https://github.com/apollographql/router/pull/4039

## 🐛 Fixes

### Bring Otel `service.name` into line with the Otel spec ([PR #4034](https://github.com/apollographql/router/pull/4034))

Handling of Otel `service.name` has been brought into line with the [Otel spec](https://opentelemetry.io/docs/concepts/sdk-configuration/general-sdk-configuration/#otel_service_name) across traces and metrics.

Service name discovery is handled in the following order:
1. `OTEL_SERVICE_NAME` env
2. `OTEL_RESOURCE_ATTRIBUTES` env
3. `router.yaml` `service_name`
4. `router.yaml` `resources` (attributes)

> If none of the above are found then the service name will be set to `unknown_service:apollo_router` or `unknown_service` if the executable name cannot be determined.

Reviewer: Ok, I guess that's probably `router` for us then?:

```suggestion
If none of the above are found then the service name will be set to `unknown_service:router` or `unknown_service` if the executable name cannot be determined.
```

By [@bnjjj](https://github.com/bnjjj) in https://github.com/apollographql/router/pull/4057

### Fix memory issues in the apollo metrics exporter ([PR #4107](https://github.com/apollographql/router/pull/4107))

There were a number of issues with the apollo metrics exporter which meant that under load the router would look as though it was leaking memory. It isn't a leak, strictly speaking, but is in fact "lingering" memory.

The root cause was a bounded `futures` channel which did not enforce the bounds as we expected and thus could over-consume memory. We have fixed the issue by:

 - making the situation of channel overflow less likely to happen
   - speeding up metrics processing
   - altering the priority of metrics submission vs metrics gathering
 - switching to use a `tokio` bounded channel which enforces the bound as we originally expected

With these changes in place we have observed that the router behaves very well with respect to memory consumption under high load.

By [@garypen](https://github.com/garypen) in https://github.com/apollographql/router/pull/4107

### Support authorization directive renaming ([PR #3949](https://github.com/apollographql/router/pull/3949))

> When importing directives through the `@link` directive, they can be renamed. This makes sure that the authorization plugin can still recognize its directives when they have been renamed.

Reviewer: ```suggestion
When importing directives into subgraph schemas using the `@link` directive it is possible to rename the directive [using the `as:` parameter](https://specs.apollo.dev/link/v1.0/#@link.as). It is now possible to do this for authorization directives and they will be recognized appropriately even if they have been renamed.
```

By [@Geal](https://github.com/Geal) in https://github.com/apollographql/router/pull/3949

## 📃 Configuration

### Bring telemetry tracing config and metrics config into alignment ([Issue #4043](https://github.com/apollographql/router/issues/4043))

> Configuration between tracing and metrics was inconsistent and did not align with otel spec terminology. The following changes have been made to router.yaml configuration:

Reviewer: ```suggestion
Configuration between tracing and metrics was inconsistent and did not align with the terminology defined in the OpenTelemetry specification. To correct this, the following changes have been made to the router's YAML configuration:
```

`telemetry.tracing.trace_config` has been renamed to `common`
  
```diff
telemetry
  tracing:
-   trace_config:
+   common:   
```

`telemetry.tracing.common.attributes` has been renamed to `resource`
```diff
telemetry
  tracing:
    common:
-      attributes:
+      resource:   
```

`telemetry.metrics.common.resources` has been renamed to `resource`
```diff
telemetry
  metrics:
    common:
-      resources:
+      resource:   
```
`telemetry.tracing.propagation.awsxray` has been renamed to `aws_xray`
```diff
telemetry
  tracing:
    propagation:
-      awsxray: true
+      aws_xray: true
```

The Router will upgrade any existing configuration on startup. However, you should update your configuration to use the new format as soon as possible. 

By [@BrynCooke](https://github.com/BrynCooke) in https://github.com/apollographql/router/pull/4044, https://github.com/apollographql/router/pull/4050 and https://github.com/apollographql/router/pull/4051

## 🛠 Maintenance

> ### Router should respond with subscription-protocol header for callback ([Issue #3929](https://github.com/apollographql/router/issues/3929))

Reviewer: ```suggestion
### Router should respond with `subscription-protocol` header for callback ([Issue #3929](https://github.com/apollographql/router/issues/3929))
```

> Callback protocol documentation specifies that router responds with `subscription-protocol: callback/1.0` header to the initialization (check) message. Currently router does not set this header on the response.

Reviewer: ```suggestion
The router will now include a `subscription-protocol: callback/1.0` header on the response to a initialization (check) message, per the [callback protocol documentation](https://www.apollographql.com/docs/router/executing-operations/subscription-callback-protocol).
```

@bnjjj Note that the [docs](https://www.apollographql.com/docs/router/executing-operations/subscription-callback-protocol) now differ from [the `dev-docs`](https://github.com/apollographql/router/blob/dev/dev-docs/callback_protocol.md)!   Can you fix that?

Reviewer: Covered by https://github.com/apollographql/router/issues/3937 already.  Thanks!

By [@Geal](https://github.com/Geal) in https://github.com/apollographql/router/pull/4076 & https://github.com/apollographql/router/pull/4122

### Add a new GraphOS Studio reporting metric ([Issue #3883](https://github.com/apollographql/router/issues/3883))

The new metric `apollo.router.telemetry.studio.reports` is a count of the number of reports the router submits to GraphOS Studio.

Its `type` attribute sets the type of the report (`traces` or `metrics`).

By [@garypen](https://github.com/garypen) in https://github.com/apollographql/router/pull/4039

## 🐛 Fixes

### Bring OTel `service.name` in line with the OTel specification ([PR #4034](https://github.com/apollographql/router/pull/4034))

Handling of OpenTelemetry (OTel) `service.name` has been brought into line with the [OTel specification](https://opentelemetry.io/docs/concepts/sdk-configuration/general-sdk-configuration/#otel_service_name) across traces and metrics.

Service name discovery is handled in the following order:
1. `OTEL_SERVICE_NAME` env
2. `OTEL_RESOURCE_ATTRIBUTES` env
3. `router.yaml` `service_name`
4. `router.yaml` `resources` (attributes)

> If none of the above are found then the service name will be set to `unknown_service:apollo_router` or `unknown_service` if the executable name cannot be determined.

Reviewer 2: ```suggestion
If none of the above are found then the service name will be set to `unknown_service:router` or `unknown_service` if the executable name cannot be determined.
```
